### DISCUSSION TITLE
Fortran interface for particles

### DISCUSSION THREAD

**Discussion by HoulinLee** [AMReX-Codes/amrex#2358] [Q&A] [answered]
Dear all,
I'm a Fortran user and trying to use AMRex in my own codes. It's mentioned in the _amrex Documentation_ that AMReX provides Fortran interfaces around most of the functionalities except for the AmrLevel class and particles. But I find the following message in the _CHANGES_ document:
_18.08 -- Fortran interface for particles._ 
I also find the example in _amrex/Tests/FortranInterface/Advection_F_ that the _amrex_particlecontainer_module_ is used. So, is there a Fortran interface for particles in the latest version of AMRex? 
Thank you.

**Comment by atmyers** [comment, marked as answer]
Hi Houlin,

That bit of documentation is out of date - there is indeed a Fortran interface for particles, and that tutorial you mentioned is currently the best place to see how it can be used. We will add some information about this to Fortran section of the documentation.

-Andrew

**Reply by HoulinLee** [reply]
Hi Andrew, thanks for your reply. And could you please tell me is there a Fortran interface for AmrLevel class now? And is there any other functionality that could not be used by Fortran? Thank you.

**Reply by atmyers** [reply]
AmrLevel is still not usable through the Fortran Interfaces, unfortunately. The most up-to-date information about what's available would be in the `amrex/Src/F_Interfaces` directory:

```
atmyers@atmyers-XPS-13-9380:~/AMReX-Codes/amrex/Src/F_Interfaces$ ls
AmrCore  Base  CMakeLists.txt  LinearSolvers  Octree  Particle
```

So in addition to particles, there are also a subset of the linear solvers available through Fortran, but functionality other than that is not available.

**Reply by HoulinLee** [reply]
OK, thanks for your help.

**Reply by HoulinLee** [reply]
Hi Andrew, I'm trying to modify my own PIC (Particle-in-Cell) codes with AMRex. But as I find in _amrex/Src/F_Interfaces/Particle/AMReX_particlecontainer_mod.F90_, it seems that there is only a subset of the particle functions available through Fortran. Is it possible to set extra particle variables (such as the particle mass) in the AMReX particle struct and interact the particle data with the mesh data in Fortran?

**Reply by atmyers** [reply]
Hi Houlin,

You will need to create your own versions of the files `amrex/Src/F_Interfaces/Particle/AMReX_particlecontainer_mod.F90` and `amrex/Src/F_Interfaces/Particle/AMReX_particlecontainer_fi.cpp` that register the correct number of components and define a Fortran-side data type for your particles.

**Reply by HoulinLee** [reply]
I created the new versions of the two files and it worked, thanks.
